Adding Extensible enum type

dsgouda opened the pull request:
Addressing https://github.com/Azure/autorest/issues/1591
Adding ExtensibleEnums type and related serialization/deserialization logic to ClientRuntime
Most of the code is derived from the solution implemented by  @brjohnstmsft 


<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
<!--
Please add an informative description that covers that changes made by the pull request.

If you are regenerating your SDK based off of a new swagger spec, please add the link to the corresponding swagger spec pull request that has been merged in the azure-rest-api-specs repository
-->

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

fearthecowboy commented:
/cc @brjohnstmsft and @mattchr

fearthecowboy commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -0,0 +1,121 @@
+                    .Select(f => f.GetValue(null))
+                    .OfType<T>();
+
+            return allKnownValues.ToDictionary(v => (string)v, v => v);
```
You probably want that to read 
```
allKnownValues.ToDictionary(v => v.ToString(), v => v);
```

The cast will fail if the type isn't explicitly coercible to `string`. `ToString()` won't fail.

brjohnstmsft replied:
`T` has a constraint that it must derive from `ExtensibleEnum<T>`, which has an explicit `operator string` implementation. That said, if the intent is to generalize this implementation from `string` to `TVal`, then this should probably be `v => v._value` (or whatever you rename `_name` to).

olydis commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -0,0 +1,121 @@
+        /// Initializes a new instance of the ExtensibleEnum class.
+        /// </summary>
+        /// <param name="name">The value for this instance of the extensible enumeration.</param>
+        protected ExtensibleEnum(string name)
```
`name` should really be `value` since it's what's about the wire here!

olydis replied:
`value` would have `TVal` and so on

dsgouda replied:
Fair point, can you think of a case where `TVal` could be a non-primitive type

olydis replied:
well, if you think about Swagger, *any* type is actually allowed. In here, I also don't (yet) see any place that depends on `TVal` being primitive, so I think this generalization should automatically allow any type

olydis commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -0,0 +1,121 @@
+    /// <summary>
+    /// Abstract base class for types that act like enums, but can be extended with arbitrary string values.
+    /// </summary>
+    public abstract class ExtensibleEnum<T> : IEquatable<T> where T : ExtensibleEnum<T>
```
let `ExtensibleEnum` take another generic argument (I'll call it `TVal`) that determines the type of stuff that the enum is about

brjohnstmsft replied:
Do you have a particular scenario in mind? Attaching labels to a fixed set of integers, for example...?

dsgouda replied:
@olydis I am trying to generalize this as you suggested but the problem is when I try to create an `ExtensibleEnum<T,V>` type I need to provide the name of the enum as well as its underlying (primitive) type. I don't think we capture this information in `EnumType.cs` so there really isn't an easy way to figure this out.
Any ideas?

olydis replied:
@brjohnstmsft exactly, supporting non-string enums is one of the bigger asks we'll probably tackle some time soon (appears in some data-plane stuff we saw coming in) 🙂

olydis replied:
@dsgouda not sure I understand, pretty sure this is expressible - what's the error you get when doing what?

olydis commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -0,0 +1,121 @@
+    /// </summary>
+    public abstract class ExtensibleEnum<T> : IEquatable<T> where T : ExtensibleEnum<T>
+    {
+        private static readonly Lazy<Dictionary<string, T>> _nameMap =
```
`string` should then be `TVal`

fearthecowboy commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -0,0 +1,121 @@
+        private static readonly Lazy<Dictionary<string, T>> _nameMap =
+            new Lazy<Dictionary<string, T>>(CreateNameMap, isThreadSafe: true);
+
+        private string _name;
```
`name` should be called `value` or `serializedValue` -- `name` is too ambigious

brjohnstmsft replied:
👍 

The name `_name` was a holdover from before I generalized this pattern into a base class. Back then it was `AnalyzerName`, which was search-specific, and really was all about names.

olydis commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -0,0 +1,121 @@
+        /// <param name="name">The value for this instance of the extensible enumeration.</param>
+        protected ExtensibleEnum(string name)
+        {
+            _name = name;
```
populate the `Dictionary` right here, also don't make it `Lazy`

dsgouda replied:
Populate the static dictionary here? Don't get you....

brjohnstmsft replied:
@olydis Any particular reason `Lazy` is not a good idea here? My intent was that code that doesn't use any of the extensible enums won't have to pay the cost of reflection for building the map.

olydis replied:
agree about the reflection, actually the plan is to get rid of that as well - as far as I'm concerned, the Dictionary should map *raw values* to *wrapped values*, i.e. the keys don't really matter. Thinking about the scenarios:

say we have enum `Color` mapping pascal-cased color names, because C# to lower case color names for the wire
- if you wanna use a "well-known" enum value, you do so using `Color.Red`
- if the color you want is not there (but you really know that the server accepts it), you gotta do `Color.Create('cyan')`, i.e. provide the wire value *anyways* (or should the Create function know that it must lowercase stuff? even worse more so if it's an int enum)

So actually, I was somewhat confused about the behavior before. Also imagine the deserializer encounters something it hasn't seen before, won't it use `Create` using the *value* to create the new wrapped instance? So even if it sees `red`, how does it know to return *the very* `Color.Red` if the Dictionary is a mapping from name to value, i.e. `Red` to the wrapped value?

brjohnstmsft replied:
@olydis 
> agree about the reflection, actually the plan is to get rid of that as well 

Did you mean get rid of the dictionary entirely, or just change how it works...? I don't think you can do either without losing the "interning" behavior.

> as far as I'm concerned, the Dictionary should map raw values to wrapped values, i.e. the keys don't really matter.

It already does this. Note the call to `GetValue` here:

```csharp
IEnumerable<T> allKnownValues = 
                allPublicStaticFields 
                    .Where(f => f.FieldType == typeof(T)) 
                    .Select(f => f.GetValue(null)) 
                    .OfType<T>(); 
```

When it gets converted to a dictionary, the key comes from the explicit `operator string`, which calls `ToString`, which returns `_name`. This is admittedly confusing... The field should have been called `_value` instead of `_name`, and the code to build the dictionary should have looked like this:

```csharp
 return allKnownValues.ToDictionary(v => v._value, v => v); 
```

But I think it already does what you want it to do.

> if the color you want is not there (but you really know that the server accepts it), you gotta do Color.Create('cyan'), i.e. provide the wire value anyways (or should the Create function know that it must lowercase stuff? even worse more so if it's an int enum)

As originally designed, you'd have to provide the wire value, which ideally would be documented in the REST API reference. `Create` is not meant to be "smart" in any way -- it just looks up exactly the key you give it, and creates a new wrapper if it isn't found. This should give the most predictable behavior.

Note that I originally had an implicit conversion from `ExtensibleEnum` to `string` (read: `TVal`), but that turned out to be very problematic in practice. Therefore, the intent is for all usage to be in terms of the `ExtensibleEnum`-derived type, unless calling code explicitly wants to deal with `TVal`s (or needs to in case of using new server features from an old client).

dsgouda replied:
@olydis any thoughts about why making the dictionary Lazy is not a good idea here?

fearthecowboy commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -0,0 +1,121 @@
+        /// </summary>
+        /// <param name="name">Name of the ExtensibleEnum value.</param>
+        /// <returns>An instance of type T with the given name, or null if no such instance could be found.</returns>
+        protected static T Lookup(string name)
```
perhaps we should change this to a static indexer -

``` csharp
protected static T this[string value] {
... 
}
```

brjohnstmsft replied:
I'm pretty sure that static indexers are [not](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/indexers/) a [thing](https://stackoverflow.com/questions/401232/static-indexers).

Also, the intent of this method is to provide derived classes with a way to implement `Create` with interning semantics. Given that it's protected, I'm not sure what the extra syntactic sugar would buy you.

dsgouda replied:
Yup, learned this the hard way

dsgouda commented:
Looks like we also have some test cases (here)[https://github.com/Azure/azure-sdk-for-net/blob/3f736b5af3851ab99bbaa98483ae537de0d48cfb/src/SDKs/Search/DataPlane/Search.Tests/Tests/ExtensibleEnumTests.cs#L10] which we may be able to leverage

brjohnstmsft commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -0,0 +1,121 @@
+        }
+
+        /// <inheritdoc />
+        public override bool Equals(object obj)
```
What's the policy on C# 6 syntax in this repo? Can we simplify these to be expression-bodied methods (`=>`)?

brjohnstmsft replied:
Nevermind, I see I already used `=>` above. I must have missed these.

brjohnstmsft commented on src/SdkCommon/ClientRuntime/ClientRuntime/Serialization/ExtensibleEnumConverter.cs:
```diff
@@ -0,0 +1,222 @@
+    using System.Reflection;
+    using Microsoft.Rest.ClientRuntime;
+    using Newtonsoft.Json;
+    /// <summary>
```
nit: Missing newline

brjohnstmsft commented on src/SdkCommon/ClientRuntime/ClientRuntime/Serialization/ExtensibleEnumConverter.cs:
```diff
@@ -0,0 +1,222 @@
+    /// <summary>
+    /// Base class for custom JsonConverters.
+    /// </summary>
+    public abstract class ConverterBase : JsonConverter
```
Maybe this should be in its own file?

brjohnstmsft commented on src/SdkCommon/ClientRuntime/ClientRuntime/Serialization/ExtensibleEnumConverter.cs:
```diff
@@ -0,0 +1,222 @@
+        /// </param>
+        public ExtensibleEnumConverter(string factoryMethodName)
+        {
+            if (factoryMethodName == null) {
```
nit: Inconsistent brace indentation.

Unsolicited sales pitch: Consider adopting the `Throw` class. It makes code like this so much shorter, and FxCop is still smart enough to see that argument checking is happening.

dsgouda replied:
Introducing newer types here may raise some concerns for @shahabhijeet  I shall leave it to his discretion

brjohnstmsft commented on src/SdkCommon/ClientRuntime/ClientRuntime/Serialization/ExtensibleEnumConverter.cs:
```diff
@@ -0,0 +1,222 @@
+
+            const string MessageFormat =
+                "No method named '{0}' could be found that is convertible to ExtensibleEnumValueFactory<{1}>.";
+
```
nit: extra line

bitsforthought reviewed (commented):
@dsgouda Thank you so much for doing this! I'm sorry I haven't had the time to do this myself (other priorities getting in the way).

brjohnstmsft commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -5,73 +5,77 @@
     {
-        private static readonly Lazy<Dictionary<string, T>> _nameMap =
-            new Lazy<Dictionary<string, T>>(CreateNameMap, isThreadSafe: true);
+        private static readonly ConcurrentDictionary<V, T> _valueMap =
```
Given that this dictionary is never mutated after it's constructed (and IMO never, ever, should be), what does using `ConcurrentDictionary` buy you?

Also, I still haven't seen a good argument against lazy initialization here.

dsgouda replied:
Agreed. @olydis thoughts?

brjohnstmsft commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -5,73 +5,77 @@
+        /// <param val="val">ExtensibleEnum to convert.</param>
         /// <returns>The ExtensibleEnum as a string.</returns>
-        public static explicit operator string(ExtensibleEnum<T> name) => name?.ToString();
+        public static explicit operator string(ExtensibleEnum<T, V> val) => val?.ToString();
```
The purpose of this conversion operator was to "unwrap" extensible enum values to their underlying types, so I guess this should be `operator V` now.

brjohnstmsft commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -5,73 +5,77 @@
+        /// <returns>An instance of type T with the given value, or null if no such instance could be found.</returns>
+
+        protected static T Lookup(V value)
+        //protected static T this[V value] 
```
nit: dead code

brjohnstmsft commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -5,73 +5,77 @@
+        //protected static T this[V value] 
         {
-            if (name == null)
+            if (value == null)
```
There is no constraint that `V` is a reference type. How does this compile?

dsgouda replied:
Hmm... I don't think I saw any compilation issues, but I guess we can check for a nullable type before comparison.

brjohnstmsft commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -5,73 +5,77 @@
+            if (value == null)
             {
-                throw new ArgumentNullException(name, "name");
+                throw new ArgumentNullException(value.ToString(), "value");
```
Calling `ToString` on something that might be null seems sketchy.

dsgouda replied:
Whoops, fixing

brjohnstmsft commented on src/SdkCommon/ClientRuntime/ClientRuntime/Serialization/EnumConverterBase.cs:
```diff
@@ -0,0 +1,127 @@
+    /// <summary>
+    /// Base class for custom JsonConverters.
+    /// </summary>
+    public abstract class EnumConverterBase : JsonConverter
```
Why the name change? There is nothing about this class that is specific to the enum converter.

Also, after staring at this with fresh eyes, I think it would have been better if I had not used inheritance for re-use purposes here. These methods could just as easily be extensions on `JsonReader`. I leave it up to you to decide what fits best with the `ClientRuntime`.

dsgouda replied:
I was trying to semantically narrow the scope of its usage since a lot of common serialization related logic lies in this project and I did not want to introduce something new.
On similar lines if we do model these methods as extension methods for the JsonReader I wonder if it may cause a breaking change (I can't think of a reason why but just want to point it out)

@shahabhijeet @fearthecowboy @olydis I'd like your thoughts too

brjohnstmsft replied:
@dsgouda Regarding backward compatibility, if you decide to make these extension methods of `JsonReader`, I would recommend doing so in an internal class so it is only visible within the ClientRuntime.

brjohnstmsft commented:
@dsgouda It would be great if you could bring in the test cases too.

dsgouda commented:
@brjohnstmsft I am working on the test cases will update the PR when ready

dsgouda commented:
@brjohnstmsft  Addressed most of your comments and added a couple of simple tests, feel free to suggest any edge cases we can test here
Still waiting on some feedback with regards to other comments

brjohnstmsft commented:
@dsgouda What about [these tests](https://github.com/Azure/azure-sdk-for-net/blob/psSdkJson6/src/SDKs/Search/DataPlane/Search.Tests/Tests/ExtensibleEnumTests.cs)? Can you include them as well?

brjohnstmsft commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -118,5 +118,17 @@ public override string ToString()
             return allKnownValues.ToDictionary(v => v._value, v => v);
         }
+
+        public static T Create(V value)
```
This introduces two problems:
  1. Now we need to reason about concurrent modification of the map. This wasn't the case before. That map should be immutable once constructed.
  1. `Activator.CreateInstance` is surely slower than just calling the constructor directly in the generated derived class. What's wrong with that way of doing it?

This doesn't seem worth the effort for what should be a rare case.

brjohnstmsft commented on src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs:
```diff
@@ -119,16 +119,6 @@ public override string ToString()
-            _valueMap.TryAdd(value, newItem);
-            return newItem;
-        }
+        protected static void CacheValue(V value, T newItem) => _valueMap.TryAdd(value, newItem);
```
My objection to modifying the map still stands. Not worth the effort relative to the cognitive load, runtime overhead, and chance of bugs that it introduces.

dsgouda replied:
I'm still unsure as to why caching newly introduced values is a bad idea. I agree it may be a bit of an overhead but I also feel it helps keep track of values created of that type (though I can't think of a concrete scenario where this is useful). I'm on the fence regarding this.

@olydis @fearthecowboy feedback would be much appreciated

olydis replied:
@brjohnstmsft what was the original motivation of the `Dictionary` in your case? I assumed it is to get referential equality as you would get it from C# enums.

E.g., when the server returns `{ "color": "red" }`, I wanna be able to say `response.Color == Color.Red` after deserialization. Then also, the extensibility of extensible enum also means that previously unseen values may be round-tripped, so not only would I expect `{ "color1": "red", "color2": "red" }` to result in `result.Color1 == result.Color2`, but also `{ "color1": "blue as the sky", "color2": "blue as the sky" }` to result in `result.Color1 == result.Color2` (where that value isn't well known). Without caching, this is not really possible, i.e. those two properties will *not* be referentially equal. But then, imagine the service adds that value to its set of known values (arguably that's exactly what we'd expect in Azure land). Suddenly the values become referentially equal? I'm scared of turning breaking changes into seemingly unrelated behavioral changes.

brjohnstmsft replied:
@dsgouda Your default instinct when considering shared mutable state should be to run away screaming. :) There should be a really, really compelling reason to have it, or else by default it's a bad idea.

@olydis The intent is to provide a small optimization for the set of known values. It's more about saving memory than about how equality is implemented. These things should have value equality semantics just like enums, regardless of how the values are constructed. If most of the time that equality check happens to use reference equality under the hood, that's fine. I'm not sure what risk you're seeing here.

olydis replied:
@brjohnstmsft You're right, I missed the `==` operator definition - think I've spent too much time in JavaScript land and figured that two objects will only compare `==` if they are truly the same. Which is obviously not the case in C#. And there's indeed no requirement in making `Object.ReferenceEquals` "work" since that doesn't work on real `enum`s either due to boxing... thanks for making me think :D

olydis replied:
so I guess the choice is:
- use caching, so one can get rid of *all* `GetHashCode`/`Equals`/`==` operator definitions
- get rid of the cache, keeping all the operators

brjohnstmsft replied:
Why isn't leaving it the way I originally designed it an option? i.e. -- cache known values to save memory and keep the current equality semantics.

brjohnstmsft replied:
@olydis Can we please close on this?

dsgouda replied:
@fearthecowboy @shahabhijeet are also taking a look into this, they should be posting their comments soon

brjohnstmsft replied:
We haven't settled on what to do with the concurrent dictionary. Just to re-raise my question above:

> Why isn't leaving it the way I originally designed it an option? i.e. -- cache known values to save memory and keep the current equality semantics.

olydis replied:
@fearthecowboy @dsgouda the decision here also affects how we are able to handle "alternative values" (do they roundtrip *and* compare equal to each other?)

brjohnstmsft replied:
@olydis I think they should, and the original implementation supported this.

shahabhijeet commented:
@azuresdkci test this please

dsgouda committed:
Adding support for Extensible enums

dsgouda committed:
Adding allowed values support

dsgouda committed:
minor

dsgouda committed:
undoing stuff

dsgouda committed:
added support for allowed values

fearthecowboy commented on src/SdkCommon/ClientRuntime/ClientRuntime.Tests/Resources/MovieTicket.cs:
```diff
@@ -0,0 +1,86 @@
+
+        public static TaxAmount Create(string value)
+        {
+            System.Console.Error.WriteLine("we're trying to deserialize "+value);
```
Probably don't want that left in there

shahabhijeet commented:
@dsgouda whenever you are ready to merge/update this PR, please reopen this PR or create a new one.
